_studio/mfx_lib/decode/vp9/src/mfx_vp9_dec_decode_hw.cpp

```diff
@@ -85,6 +85,164 @@ bool CheckHardwareSupport(VideoCORE *p_core, mfxVideoParam *p_video_param)
+
+            // The next condition is for decoder robustness.
+            // After the first keyframe is decoded this frame occupies all ref slots
+            // its mask is 011111111 and the condition "minus one" never hits.
+            // But if the first frame is Inter-frame encoded in the intra-only mode
+            // then we can receive "minus one" ref_frame_map item.
+            // The decoder never should touch these references.
+            // But if the stream is broken then decode may touch not initialized reference
+            // So to prevent crash let's put something as the reference frame.
+            || (m_frameInfo.ref_frame_map[ref_index] == (UMC::FrameMemID)-1))
+        {
+            MFX_CHECK((m_FrameAllocator->IncreaseReference(m_frameInfo.currFrame) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+
+            if (m_frameInfo.ref_frame_map[ref_index] >= 0)
+                MFX_CHECK((m_FrameAllocator->DecreaseReference(m_frameInfo.ref_frame_map[ref_index]) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+            m_frameInfo.ref_frame_map[ref_index] = m_frameInfo.currFrame;
+
+            m_frameInfo.sizesOfRefFrame[ref_index].width = m_frameInfo.width;
+            m_frameInfo.sizesOfRefFrame[ref_index].height = m_frameInfo.height;
+        }
+    }
+    return MFX_ERR_NONE;
+}
+mfxStatus VideoDECODEVP9_HW::CleanRefList()
+{
+    for (mfxI32 ref_index = 0; ref_index < NUM_REF_FRAMES; ++ref_index)
+    {
+        if (m_frameInfo.ref_frame_map[ref_index] >= 0)
+            MFX_CHECK((m_FrameAllocator->DecreaseReference(m_frameInfo.ref_frame_map[ref_index]) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+
+        m_frameInfo.ref_frame_map[ref_index] = -1;
+    }
+    return MFX_ERR_NONE;
+}
+
+
+class FrameStorage
+{
+private:
+    std::vector<UMC_VP9_DECODER::VP9DecoderFrame> m_submittedFrames;
+    mfx_UMC_FrameAllocator *m_frameAllocator;
+
+    void LockResources(const UMC_VP9_DECODER::VP9DecoderFrame & frame) const
+    {
+        // lock current frame for decode
+        VP9_CHECK_AND_THROW((m_frameAllocator->IncreaseReference(frame.currFrame) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+        // lock frame for copy
+        if (frame.show_existing_frame)
+        {
+            VP9_CHECK_AND_THROW((m_frameAllocator->IncreaseReference(frame.ref_frame_map[frame.frame_to_show]) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+        }
+        // lock all references
+        else
+        {
+            for (mfxI32 ref_index = 0; ref_index < NUM_REF_FRAMES; ++ref_index)
+            {
+                const UMC::FrameMemID mid = frame.ref_frame_map[ref_index];
+                if (mid >= 0)
+                {
+                     VP9_CHECK_AND_THROW((m_frameAllocator->IncreaseReference(mid) == UMC::UMC_OK), MFX_ERR_UNKNOWN);
+                }
+            }
+        }
+    }
+
+    void UnLockResources(const UMC_VP9_DECODER::VP9DecoderFrame & frame) const
+    {
+        // decoded frame should be unlocked in async routine
```

Comment:
I believe we have to unlock decoded frame here instead of doing that at async routine. What is a reason of not doing that?